### PR TITLE
Make PDF generation async

### PR DESCRIPTION
**Pull request by stefanv**
- Execute the PDF generation in a thread
- Download the file in chunks

### CODE REVIEW

**Review by acrellin — APPROVED** [review]

File: skyportal/handlers/api/source.py [hunk 1]
```diff
@@ -763,4 +770,26 @@ def get(self, obj_id):
             'Cache-Control', 'no-store, no-cache, must-revalidate, max-age=0'
         )
 
-        return self.write(image)
+        # From
+        # https://bhch.github.io/posts/2017/12/serving-large-files-with-tornado-safely-without-blocking/
+        chunk_size = 1024 * 1024 * 1  # 1 MiB
+        while True:
```
Comment (profjsb) [inline skyportal/handlers/api/source.py:776]: Seems like we should have a timeout check. Like if it takes more than X seconds (X ~ 60) then throw a notification and break. As written it's not clear if this loop cannot go forever ...

**Review by profjsb — COMMENTED** [review]
Looks good. Comments inline. Perhaps send a front end notification saying that "Finding chart generation has begun. The PDF will download automatically once it's complete."

File: skyportal/handlers/api/source.py (hunk 1 above)
Comment (stefanv) [reply skyportal/handlers/api/source.py:776]: The loop terminates in one of two ways:

1. No more bytes left in buffer
2. Stream is closed

We always try to write inside the loop, so if that fails we're out.  Other exceptions will bubble through, so I can keep an eye on the logs to see when that happens.

Comment (stefanv) [reply skyportal/handlers/api/source.py:776]: This is also a stopgap measure, until I can get the caching infrastructure fixed up—then we can serve up directly via nginx.

**Comment by stefanv** [comment]
@profjsb I've added the front-end notification; thanks, that was a good idea.

File: skyportal/handlers/api/source.py (hunk 1 above)
Comment (stefanv) [reply skyportal/handlers/api/source.py]: OK, I've now made this a for-loop, predicated on the max file size we want to be able to send.

File: skyportal/handlers/api/source.py [hunk 2]
```diff
@@ -752,8 +758,23 @@ def get(self, obj_id):
...
+
         filename = rez["name"]
-        image = rez["data"]
+        image = io.BytesIO(rez["data"])
+
+        # Adapted from
+        # https://bhch.github.io/posts/2017/12/serving-large-files-with-tornado-safely-without-blocking/
+        mb = 1024 * 1024 * 1
+        chunk_size = 1 * mb
+        max_file_size = 5 * mb
```
Comment (profjsb) [inline skyportal/handlers/api/source.py:773]: These can be pretty large if someone chooses a large image size window in the finder. I'd make this max size = 15 Mb

Comment (stefanv) [reply skyportal/handlers/api/source.py:773]: Sure, np

File: skyportal/handlers/api/source.py [hunk 3]
```diff
@@ -752,8 +758,23 @@ def get(self, obj_id):
...
+        if not (image.getbuffer().nbytes < max_file_size):
+            return self.error(
+                f"Refusing to send files larger than {max_file_size / mb} MB"
```
Comment (profjsb) [inline skyportal/handlers/api/source.py:776]: Maybe clip this to 1 sig fig after the decimal place

Comment (stefanv) [reply skyportal/handlers/api/source.py]: Addressed.

File: skyportal/handlers/api/source.py (hunk 2 above)
Comment (stefanv) [reply skyportal/handlers/api/source.py]: Done

**Review by profjsb — APPROVED** [review]

**Pull request merged**